### PR TITLE
Apply type depth limit to :new results

### LINKED ISSUE #14009: Avoid infinite type recursion in type inference
See point 2 in https://github.com/JuliaLang/julia/issues/14005#issuecomment-156849772

A [minimum repro](https://github.com/yuyichao/explore/blob/9f1f81b453ee3c9b70a09e8e12c3bd81a990cb82/julia/wb-freeze/typeinf-recursion.jl) is,

```jl
type A{T}
end

A{T}(a::T) = A{T}()

function f(a)
    if rand() > 100
        return f(A(a))
    end
end

@code_warntype f(1)
```

Even though the recursion path will never be executed, type inference will still try to infer the type of the recursion call with a infinitely deep type and even if there's a genuine infinite recursion, it's better to do that at runtime and not in type inference.....

We already have a bunch of limit on type length so I think we should probably also have some limit here.

@carnaval Does green-fairy provide any general solution to these problems?

### PR DESCRIPTION
**Pull request by carnaval**
Should fix #14009.

### CODE REVIEW

**Comment by yuyichao** [comment]
:+1: =)

File: test/core.jl [hunk 1]
```diff
@@ -3499,3 +3499,10 @@ let
     finalize(obj)
     @test finalized == 1
 end
+
+# non-terminating inference, issue #14009
+type A14009{T}
+end
+A14009{T}(a::T) = A14009{T}()
+f14009(a) = ? ? f14009(A14009(a)) : a
```
Comment (yuyichao) [inline test/core.jl]: Any less confusing way of specifying a condition that type inference cannot reason about?

Comment (carnaval) [inline test/core.jl]: sure, I happen to really like this but I think I'm alone on this one. I somewhat remember convincing Jeff that it made question-mark-as-a-global a feature but maybe he was just being nice.

**Comment by JeffBezanson** [comment]
I'm not sure this is a good way to fix this, since `new` itself is not responsible for any growth in the type. This will penalize all code that might allocate deeply-nested types, even without recursion.

We should try to do the limiting during function calls, perhaps in abstract_call, and ideally only *if* we detect recursion might be happening.

**Comment by carnaval** [comment]
Agreed, but I think the quick fix in the PR is analogous to how we do the tuple depth limit
```
julia> f() = ((((((1,),),),),),)
julia> Base.return_types(f,())[1]
Tuple{Tuple{Tuple{Tuple{Tuple{Tuple}}}}}
```
since tuple() and new() play the same role ?

**Comment by JeffBezanson** [comment]
Yes, if we want a quick fix then I'd much rather do this in `apply_type` than in `new`.

**Comment by JeffBezanson** [comment]
I will submit a different version of this.

**Pull request closed without merging**